text/006-local-js-dependencies.md

```diff
@@ -0,0 +1,280 @@
+now, while paving a way forward to eventually support local JS snippets in
+Node.js. The `--no-modules` could eventually also be supported in the same
+manner as Node.js is (once we're parsing the JS file and rewriting the exports),
+but it's proposed here to generally move away from `--no-modules` towards
+`--browser`.
+
+For some more detail about the `--browser` output, it's intended to look from
+the outside like `--no-modules` does today. When using `--browser` a single
+`wasm_bindgen` function will be exported from the module. This function takes
+either a path to the wasm file or the `WebAssembly.Module` itself, and then it
+returns a promise which resolves to a JS object that has the full wasm-bindgen
+interface on it.
+
+### JS files depending on other JS files
+
+One tricky point about this RFC is when a local JS snippet depends on other JS
+files. For example your JS might look like:
+
+```js
+// js/foo.js
+
+import { foo } from '@some/npm-package';
+import { bar } from './bar.js'
+
+// ...
+```
+
+As designed above, these imports would not work. It's intended that we
+explicitly say this is an initial limitation of this design. We won't support
+imports between JS snippets just yet, but we should eventually be able to do so.
+
+In the long run to support `--nodejs` we'll need some level of ES module parser
+for JS. Once we can parse the imports themselves it would be relatively
+straightforward for `#[wasm_bindgen]`, during expansion, to load transitively
+included files. For example in the file above we'd include `./bar.js` into the
+wasm custom section. In this future world we'd just rewrite `./bar.js` (if
+necessary) when the final output artifact is emitted. Additionally with NPM
+package support in `wasm-pack` and `wasm-bindgen` (a future goal) we could
+automatically add entries to `package.json` (or validate they're already
+present) based on the imports found.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* The initial RFC is fairly conservative. It doesn't work with `--nodejs` out of
+  the gate nor `--no-modules`. Additionally it doesn't support JS snippets
+  importing other JS initially. Note that all of these are intended to be
+  supported in the future, it's just thought that it may take more design than
+  we need at the get-go for now.
+
+* JS snippets must be written in vanilla ES module JS syntax. Common
+  preprocessors like TypeScript can't be used. It's unclear how such
+  preprocessed JS would be imported. It's hoped that JS snippets are small
+  enough that this isn't too much of a problem. Larger JS snippets can always be
+  extracted to an NPM package and postprocessed there.
+
+* The relatively popular `--no-modules` flag is proposed to be deprecated in
+  favor of a `--browser` flag, which itself will have a breaking change relative
+  to today. It's thought though that `--browser` is only very rarely used so is
+  safe to break, and it's also thought that we'll want to avoid breaking
```

Comment:
I do have concerns about changing the meaning of existing flags and whether or not we would need to do a breaking bump. I am not sure that this is "safe" (I have no idea and no data and I don't think you have any more data than I have!)

Reply:
Oh sorry meant to respond to this earlier, but you're spot on that I have very little data about this :)

I'm curious to hear what others think about this, although I suspect you're right and that we'll need to make new flags instead of trying to repurpose existing flags.

Reply:
This PR seems primarily focused on module resolution/loaders, so perhaps their names could reflect this better? For example, we could have something like `--module-resolution={none,node,browser}`:
- `none` is the default suggested in this PR and means that either no modules need to be loaded or some bundler is expected
- `node` and `browser` replace the flags suggested in this PR

There is [some precedent in TypeScript](https://www.typescriptlang.org/docs/handbook/module-resolution.html) too.

Reply:
Hm the module resolution aspect looks like it's more worried about how to resolve the paths in file dependencies (aka the `"foo"` in `import x from "foo"`). For us though it's more about what to generate at all, e.g. `import` or `require` or "somehow nothing"

Reply:
You're right, the concepts are fairly related though (especially with mjs in Node). Maybe `module-loader` or similar instead of `module-resolution`?

`browser` and `node` seem a bit general because they could imply whether web APIs etc. are available as well. Although that might be ok if the flags will eventually be used beyond module concerns.

Reply:
They're similar yeah, but the flags proposed here are intended to capture the use cases that basically everyone seems to fall into, which is "browsers with bundlers", "browsers without bundlers", and "node"